core/src/storage/seed_encryption.py

```diff
@@ -0,0 +1,217 @@
+from typing import List, Tuple
+
+from storage import cache_codec, cache_common
+from trezor import utils
+
+FIELDS_TO_ENCRYPT = [cache_common.APP_COMMON_SEED]
+if not utils.BITCOIN_ONLY:
+    FIELDS_TO_ENCRYPT += [
+        cache_common.APP_CARDANO_ICARUS_SECRET,
+        cache_common.APP_CARDANO_ICARUS_TREZOR_SECRET,
+    ]
+
+FIELDS_TO_ENCRYPT_SESSIONLESS = [cache_common.APP_COMMON_SEED_WITHOUT_PASSPHRASE]
+
+
+def is_empty(session: cache_codec.DataCache, fields_to_encrypt: List[int]) -> bool:
+    """
+    Checks if the session has no data set for the fields to encrypt.
+    """
+    for field in fields_to_encrypt:
+        if session.get(field):
+            return False
+    return True
+
+
+def get_seed_encryption_key(session_id: bytes) -> bytes:
+    """
+    Returns the seed encryption key for a given session ID.
+    The key is derived from the device secret and the session ID.
+    """
+    from typing import TYPE_CHECKING
+
+    from storage.device import get_device_secret
+
+    from apps.common.seed import Slip21Node
```

Comment:
we absolutely can't import from the `apps` namespace in `storage`

if we need the slip21 node impl here, I suggest moving it into `trezor.crypto`; that will likely involve a bigger refactor though, we may not want to just dump `apps.common.seed` in there as a whole.
perhaps just take out the implementation of the derivation process as a simple function?